src/dex/infusion/infusion.ts

```diff
@@ -0,0 +1,695 @@
+
+  constructor(
+    protected network: Network,
+    dexKey: string,
+    protected dexHelper: IDexHelper,
+    isDynamicFees = true,
+    factoryAddress?: Address,
+    subgraphURL?: string,
+    initCode?: string,
+    feeCode?: number,
+    poolGasCost?: number,
+    routerAddress?: Address,
+  ) {
+    super(
+      network,
+      dexKey,
+      dexHelper,
+      isDynamicFees,
+      factoryAddress !== undefined
+        ? factoryAddress
+        : InfusionConfig[dexKey][network].factoryAddress,
+      subgraphURL === ''
+        ? undefined
+        : subgraphURL !== undefined
+        ? subgraphURL
+        : InfusionConfig[dexKey][network].subgraphURL,
+      initCode !== undefined
+        ? initCode
+        : InfusionConfig[dexKey][network].initCode,
+      feeCode !== undefined ? feeCode : InfusionConfig[dexKey][network].feeCode,
+      poolGasCost !== undefined
+        ? poolGasCost
+        : InfusionConfig[dexKey][network].poolGasCost,
+      infusionPairIface,
+      Adapters[network] || undefined,
+    );
+
+    this.stableFee = InfusionConfig[dexKey][network].stableFee;
+    this.volatileFee = InfusionConfig[dexKey][network].volatileFee;
+
+    this.factory = new dexHelper.web3Provider.eth.Contract(
+      infusionFactoryABI as any,
+      factoryAddress !== undefined
+        ? factoryAddress
+        : InfusionConfig[dexKey][network].factoryAddress,
+    );
+
+    this.router =
+      routerAddress !== undefined
+        ? routerAddress
+        : InfusionConfig[dexKey][network].router || '';
+
+    this.feeFactor =
+      InfusionConfig[dexKey][network].feeFactor || this.feeFactor;
+
+    // this.exchangeRouterInterface = new Interface(infusionRouterABI);
+  }
+
+  async findInfusionPair(from: Token, to: Token, stable: boolean) {
+    if (from.address.toLowerCase() === to.address.toLowerCase()) return null;
+    const [token0, token1] =
+      from.address.toLowerCase() < to.address.toLowerCase()
+        ? [from, to]
+        : [to, from];
+
+    const typePostfix = this.poolPostfix(stable);
+    const key = `${token0.address.toLowerCase()}-${token1.address.toLowerCase()}-${typePostfix}`;
+    let pair = this.pairs[key];
+    if (pair) return pair;
+
+    let exchange = await this.factory.methods
+      // Infusion has additional boolean parameter "StablePool"
+      // At first we look for uniswap-like volatile pool
+      .getPair(token0.address, token1.address, stable)
+      .call();
+
+    if (exchange === NULL_ADDRESS) {
+      pair = { token0, token1, stable };
+    } else {
+      pair = { token0, token1, exchange, stable };
+    }
+    this.pairs[key] = pair;
+    return pair;
+  }
+
+  async batchCatchUpPairs(pairs: [Token, Token][], blockNumber: number) {
+    if (!blockNumber) return;
+    const pairsToFetch: InfusionPair[] = [];
+    for (const _pair of pairs) {
+      for (const stable of [false, true]) {
+        const pair = await this.findInfusionPair(_pair[0], _pair[1], stable);
```

Comment:
Could be improved with `await Promise.all([...])`

Reply:
done